hibernate-core/src/main/java/org/hibernate/query/sqm/tree/expression/JpaCriteriaParameter.java

```diff
@@ -27,25 +27,22 @@
  * @author Steve Ebersole
  */
 public class JpaCriteriaParameter<T>
-		extends AbstractSqmExpression<T>
-		implements SqmParameter<T>, QueryParameterImplementor<T> {
+		extends AbstractSqmParameter<T>
```

Comment:
This change makes me a bit nervous. You're doing this to inherit the `applyInferableType` implementation that allows changing the type of a parameter, but in the JPA Criteria API, the type of a parameter ought to be "fixed" AFAIU. Maybe it's better to add a field `hibernateType` to `JpaCriteriaParameter` that stores the type inference result? Or try to avoid this mutability in the first place and try to enhance the code part that does the type checking?

Reply:
You're right, this should not affect "regular" JPA Criteria parameters, i.e. created with `CriteriaBuilder#parameter`, which should be immutable.

This is only about parameters created implicitly with the `#set( Path<Y>, X extends Y value )` method, I think we should actually infer the parameter type right when it's created - so we do some type-checking early and possibly even coerce the value to the right type - and we have no trouble when eventually asserting assignability.